src/lib/device-logs/backends/redis.ts

```diff
@@ -76,19 +76,25 @@ export class RedisBackend implements DeviceLogsBackend {
 		const redisLogs = logs.map(this.toRedisLog, this);
 
 		const limit = ctx.retention_limit || 0;
-		const key = this.getKey(ctx);
+		const logKey = this.getKey(ctx);
+		const logBytesWrittenKey = this.getLogBytesWrittenKey(ctx);
 		// Create a Redis transaction
 		const tx = this.cmds.multi();
 		// Add the logs to the List structure
-		tx.rpush(key, redisLogs);
+		tx.rpush(logKey, redisLogs);
 		// Trim it to the retention limit
-		tx.ltrim(key, -limit, -1);
+		tx.ltrim(logKey, -limit, -1);
 		// Publish each log using Redis PubSub
+		let logBytesWritten = 0;
 		for (const rLog of redisLogs) {
-			tx.publish(key, rLog);
+			tx.publish(logKey, rLog);
+			logBytesWritten += rLog.length;
 		}
+		// Increment log bytes written total
+		tx.incrby(logBytesWrittenKey, logBytesWritten);
 		// Devices with no new logs eventually expire
-		tx.pexpire(key, KEY_EXPIRATION);
+		tx.pexpire(logKey, KEY_EXPIRATION);
+		tx.pexpire(logBytesWrittenKey, KEY_EXPIRATION);
```

Comment:
both keys expire together (they represent the same device)

Reply:
Why do we expire the data about logs written? It's not an issue to do so imo but I'm wondering if there's a reason we want to

Reply:
I also wonder if this will cause an issue where it looks like a device suddenly has a massive negative delta?

Reply:
A1: If a device is no longer logging, we can safely expire the tally of bytes it has written as well (a counter associated with a process (logs writing) which no longer exists)

Reply:
A2: deltas are computed based on this expression:

```
for (const device of Object.keys(nowSample.data)) {
	const deltaBytes = nowSample.data[device] - lastSample.data?.[device] ?? 0;
	const bytesPerMin = 60 * (deltaBytes / (msSince(lastSample.t) / 1e3));
	metrics.histogram('device_logs_bytes_written_per_min_hist', bytesPerMin);
}
```

Where the `nowSample` consists of data from `device:*:logBytesWritten` keys which are *currently* present (sampled "just now"). So if we sampled (and computed the delta) from a certain amount (say, `x`) of logBytesWritten last minute (eg. there is a `device:D3ADB33F:logBytesWritten: x` in `lastSample.data` map), but this minute there is no longer an entry (ie. an expiry time less than a minute, something we will never see in practice), it would still not result in a negative delta since `device:D3ADB33F:logBytesWritten` is not a key of `nowSample.data`.

([source link in context of `sampler.ts`](https://github.com/balena-io/balena-redis-exporter/blob/first-build/device-logs-bytes-written-exporter/src/sampler.ts#L76))